test/e2e/network/service.go

```diff
@@ -1682,10 +1682,35 @@ var _ = common.SIGDescribe("Services", func() {
 
 		service := t.BuildServiceSpec()
 		service.Spec.Type = v1.ServiceTypeNodePort
-
+		errAllocated := errors.New("provided port is already allocated")
+		noNodePortAvailable := errors.New("no available nodeport found")
 		ginkgo.By("creating service " + serviceName + " with type NodePort in namespace " + ns)
-		service, err := t.CreateService(service)
+		err := retry.OnError(retry.DefaultRetry, func(e error) bool {
+			if e == nil {
+				return false
+			}
+			if strings.Contains(e.Error(), errAllocated.Error()) ||
+				strings.Contains(e.Error(), noNodePortAvailable.Error()) {
+				return true
+			}
+			return false
+		}, func() error {
+			v, ok := e2eservice.GetUnusedStaticNodePort()
+			if !ok {
+				return noNodePortAvailable
+			}
+			service.Spec.Ports[0].NodePort = int32(v)
+			var err error
+			service, err = t.CreateService(service)
+			return err
+		})
 		framework.ExpectNoError(err, "failed to create service: %s in namespace: %s", serviceName, ns)
+		nodePort := service.Spec.Ports[0].NodePort
+		ok := e2eservice.ReserveStaticNodePort(int(nodePort))
+		if !ok {
+			framework.Logf("Static node port allocator was not able to reserve nodeport: %d", nodePort)
+		}
+		defer e2eservice.ReleaseStaticNodePort(int(nodePort))
```

Comment:
I think we can simplify this and make it easy to debug, this retry helpers are very obscure and if we fail more than 5 times to create a port because there is a conflict we have a problem
```suggestion
    // try to allocate a nodePort that is free
    for i:=0; i< 5; i++{
        	port, ok := e2eservice.GetUnusedStaticNodePort()
	        if !ok {
			framework.Logf("Static node port allocator was not able to reserve nodeport so we use the one assigned by the apiserver")
		} else {
		    service.Spec.Ports[0].NodePort = int32(port)
		}
		service, err = t.CreateService(service)
		if err == nil {
		      nodePort := service.Spec.Ports[0].NodePort
		      	ok := e2eservice.ReserveStaticNodePort(int(nodePort))
		        if !ok {
			    framework.Logf("Static node port allocator was not able to reserve nodeport: %d", nodePort)
           		}
		      break
	        }
                if apierrors.IsConflict(err) {
                        framework.Logf("node port %d is already allocated to other service, retrying ... : %v", nodePort, err)                  continue          
                }
                framework.ExpectNoError(err, "failed to create service: %s in namespace: %s", serviceName, ns)
                              
    }
```